src/hooks/useMapState/useMapOrObjectState.ts

```diff
@@ -0,0 +1,22 @@
+import { usePrevious } from "../usePrevious";
+import type { MapControls } from "./useMapState";
+import { useMapState } from "./useMapState";
+import type { UseObjectStateResult } from "./useObjectState";
+import { useObjectState } from "./useObjectState";
+
+export function useMapOrObjectState<K, V>(
+  initialValue: Map<K, V>
+): MapControls<K, V>;
+export function useMapOrObjectState(initialValue: object): UseObjectStateResult;
+export function useMapOrObjectState(
+  initialValue: Map<any, any> | object
+): MapControls<any, any> | UseObjectStateResult {
+  const isMap = initialValue instanceof Map;
+  const wasMap = usePrevious(isMap);
+  if (wasMap !== null && isMap !== wasMap)
+    throw new Error("Cannot switch from using map to object or vice versa");
+  /* eslint-disable react-hooks/rules-of-hooks */
+  if (isMap) return useMapState(initialValue as Map<any, any>);
+  else return useObjectState(initialValue);
+  /* eslint-enable react-hooks/rules-of-hooks */
+}
```

Comment:
>  the only thing is that it only accepts Map type as `initalValue`.

@qiweiii  It also accepts an object, and uses the old useMapState if an object is given. So it doesn't have any breaking changes.

Reply:
Oh I see, it's all good then.